File: week2/csharp/main/repository/AbstractRepository.cs
using System.Collections.ObjectModel;
using csharp.model;

namespace csharp.repository
{
    public class AbstractRepository<T, String> : IRepository<T, String> where T : IIdentifiable<String> where String : notnull
    {
        protected Dictionary<String, T> elements = new();

        public void Add(T item)
        {
            if (elements.ContainsKey(item.getId()))
            {
                throw new Exception("Element already exists!");
            }

            elements[item.getId()] = item;
        }

        public void Delete(T item)
        {
            elements.Remove(item.getId());
        }

        public IEnumerable<T> FindAll()
        {
            return elements.Values.ToList();
        }

        public T FindById(String id)
        {
            if (!elements.ContainsKey(id))
            {
                throw new Exception("Element does not exist!");
            }

            return elements[id];
        }

        public void Update(T item, String id)
        {
            if (!elements.ContainsKey(id))
            {
                throw new Exception("Element does not exist!");
            }

            elements[id] = item;
        }

        public IList<T> GetAll()
        {
            return elements.Values.ToList();
        }
    }
}

File: week2/gradle-java/src/main/java/repository/AbstractRepository.java
package repository;

import model.Identifiable;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public class AbstractRepository <T extends Identifiable<String>, String> implements Repository<T, String> {
    protected Map<String, T> elements;

    public AbstractRepository(){
        elements = new HashMap<>();

    }
    public void add(T el){
        if(elements.containsKey(el.getId()))
        {
            throw new RuntimeException("Element already exists!");
        }
        else
            elements.put(el.getId(),el);
    }

    public void delete(T el){
        elements.remove(el.getId());
    }

    @Override
    public void update(T elem, String id) {
        if(elements.containsKey(id))
            elements.put(elem.getId(),elem);
        else
            throw new RuntimeException("Element does not exist");
    }

    @Override
    public T findByBarCode(String id) {
        if(elements.containsKey(id))
            return elements.get(id);
        else
            throw new RuntimeException("Element does not exist");
    }

    public Iterable<T> findAll() {
        return elements.values();
    }

    @Override
    public Collection<T> getAll() {
        return elements.values();
    }
}
